src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -198,86 +88,204 @@ object LoadMatrix {
-        if (line.nonEmpty) {
-          var newoff = line.indexOf(sep)
-          if (newoff == -1)
-            newoff = line.length
-          val k = line.substring(0, newoff)
-          region.clear()
-          rvb.start(keyType)
-          rvb.startStruct()
-          rvb.addString(k)
-          rvb.addString(k)
-          rvb.endStruct()
-          rv.setOffset(rvb.end())
-        }
-        rv
-      }
-    }
 
     val rdd = lines.filter(l => l.value.nonEmpty)
       .mapPartitionsWithIndex { (i, it) =>
         val region = Region()
         val rvb = new RegionValueBuilder(region)
         val rv = RegionValue(region)
 
-        if (firstPartitions(i))
-          it.next()
+        if (firstPartitions(i)) {
+          val hd1 = header1Bc.value
+          val (annotationNamesCheck, hd) = parseHeader(it.next().value, seperator, nAnnotations, annotationHeaders)
+          if (!annotationNames.sameElements(annotationNamesCheck)) {
+            fatal("column headers for annotations must be the same accross files.")
+          }
+          if (!hd1.sameElements(hd)) {
+            if (hd1.length != hd.length) {
+              fatal(
+                s"""invalid sample ids: lengths of headers differ.
+                   |    ${ hd1.length } elements in ${ files(0) }
+                   |    ${ hd.length } elements in ${ fileByPartition(i) }
+               """.stripMargin
+              )
+            }
+            hd1.zip(hd).zipWithIndex.foreach { case ((s1, s2), j) =>
+              if (s1 != s2) {
+                fatal(
+                  s"""invalid sample ids: expected sample ids to be identical for all inputs. Found different sample ids at position $j.
+                     |    ${ files(0) }: $s1
+                     |    ${ fileByPartition(i) }: $s2""".
+                    stripMargin)
+              }
+            }
+          }
+        }
+
+        val at = matrixType.vaType.asInstanceOf[TStruct]
 
         it.map { v =>
           val line = v.value
-          var newoff = line.indexOf(sep)
-          if (newoff == -1)
-            newoff = line.length
-          val rowKey = line.substring(0, newoff)
+          var off = 0
+          var missing = false
+
+          def getString(file: String, rowID: Annotation, colNum: Int): String = {
+            var newoff = line.indexOf(sep, off)
+            if (newoff == -1) {
+              newoff = line.length
+            }
+            val v = line.substring(off, newoff)
+            off = newoff + 1
+            if (v == missingValue){
+              missing = true
+              null
+            } else v
+          }
+
+          def getInt(file: String, rowID: Annotation, colNum: Int): Int = {
+            var newoff = off
+            var v = 0
+            var isNegative = false
+            if (line(off) == sep) {
+              fatal(s"Error parsing matrix. Invalid Int32 at column: $colNum, row: ${ if (rowID == null) "null" else rowID.toString() } in file: $file")
```

Comment:
You can use Type.str() to print a (possibly null) annotation.  It prints null as NA.